### PR TITLE
Optimized screengrabbing

### PR DESCRIPTION
**Pull request by Werseter**
My approach to screengrabbing, based on multiple pull requests and own experience, which allowed about 40x times the FPS on a laptop at top times. Results can be flawed by multiple pulls of the exact same frame after a certain threshold. Can be checked using *allclose* from numpy if needed. Successfully used in multiple projects.

EDIT: As found out in crosstesting, whereas the grabscreen optimization was small, however noticable, the threading approach, while improving responsiveness in frame-to-frame image analysis, won't improve the efficiency in neural network approach - a further advancement into multi-core processing will be made in future.

### CODE REVIEW

**Comment by renderedSafe** [comment]
Awesome, I'm really glad you made a pull request with this. So I'm looking through the code and trying to figure out how to interface this with a collect_data type script. I'm fairly unfamiliar with threading, but as far as I can tell, I should be initializing a `getFrameThread` object, running its `start()` method, and on each iteration of the  collect_data while loop that will be looking at the screen and user input, calling `returnFrame()`, which returns `self.printscreen.getFrame()`. Does that sound about right? I really appreciate that you put this together, I may do some cleanup and documentation and submit a pull request to your repo on it, if that's ok. =)

**Comment by renderedSafe** [comment]
Also, how do you recommend resizing the image this returns?

**Comment by Werseter** [comment]
You begin with creating the threadObject 
`frameThread = getFrameThread(x1, x2, y1, y2, window_title_substring).start()` - it initializes it as an already started instance
You get a frame from the object with
`frameThread.returnFrame()`

You can try comparing the *scale* part between my and dandrews code, I think I cut it out, or use cv2 for it for resizing.

**Comment by renderedSafe** [comment]
@Werseter 
Awesome, I'm going to test this out. I'm almost done adding some docstrings for personal use, then I might convert to PEP8 and submit a pull request for ya. Once again, thanks for making this, I'll get back with how it works out for me.

**Comment by renderedSafe** [comment]
So here you go, I was running it briefly to see if it even worked for me, and I saw the "loop took X seconds" prints running by seemingly faster than usual. I had it resizing with cv2 to 480 by 270, and by the time that I saw something was happening, it had saved 2 training_data.npy files full of 500 data samples each. I just did a calculation on one of the "loop took X seconds" prints, and it comes out to about 1000 frames per second, which I believe is outrunning the thread's looping limit of about 80 FPS, so it seems that I'll have to do a similar check in the collect_data script so I don't get a bunch of duplicate images saved (also so I don't fill my HDD in a couple minutes). Very good work, I'm about to test it in GTA. I'll submit a pull request to you real quick with some cleanups.
image test
run screeny

**Comment by Werseter** [comment]
The *while True* can give a lot of false positives without checking, try adjusting the time *if* parameter inside it in getFrame.py

**Comment by renderedSafe** [comment]
Here's some cleanup and commenting


[Werseter.zip](https://github.com/Sentdex/pygta5/files/1593890/Werseter.zip)

**Comment by Werseter** [comment]
I do omit the eqality checks as my CPU runs quite slow with both GTA and python running.

**Comment by renderedSafe** [comment]
@Werseter I understand that you're trying to limit the FPS in the thread, but my ungodly FPS was coming from that fact that I was not limiting the loop in collect_data, which was not limited by anything really, and it was just calling whatever the thread object had in its memory for self.image at every iteration, so I'm just going to slow down my loop in collect_data.py to about match the iteration frequency of the thread, or maybe slightly below. I may end up changing the threading frequency by adjusting that time check in the if, but I'm going to run GTA real quick and see how that goes.

**Comment by Werseter** [comment]
That's a good call. I've checked your comments - you don't need the 40px offset with this method anymore.

**Comment by Werseter** [comment]
I also haven't mentioned it yet, as there were some uncertainties in the previous method - I've confirmed this works for all window modes - windowed, borderless and fullscreen - at least in my apps.

**Comment by renderedSafe** [comment]
So I ran it again, and actually queried the get_fps method from the grab
frame object, and even after removing the time check comparison if all
together, the thread is maxing out at 22 fps

On Dec 29, 2017 3:30 PM, "Werseter" <notifications@github.com> wrote:

That's a good call. I've checked your comments - you don't need the 40px
offset with this method anymore.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/Sentdex/pygta5/pull/84#issuecomment-354502694>, or mute
the thread

**Comment by Werseter** [comment]
I suppose the improvement might depend on the use - I've ditched the neural approach and went on further with the idea of frame-to-frame analysis, as my laptop can't handle it.

**Comment by renderedSafe** [comment]
Yeah so it seems that the thread is updating that often but whatever loop
you're using to access the self.image attribute may run a lot faster and
grab the same frame a few times, although that just made me think that
frame to frame at 20 fps might look very similar, so even if you get a
duplicate screenshot, it would likely look very similar to the frame
directly before it, and the controls you're inputting would be updated at a
much higher rate and attributed to the same image, so you may effectively
get useful data, even though you're labeling duplicate images with
different inputs.

**Comment by renderedSafe** [comment]
When I get back to my computer I'm going to try changing the code to
incorporate multiple threads each capturing a slice of the screen

**Comment by Werseter** [comment]
The issue with grabbing the exact frame can be solved with properly implemented ID for every frame generated - I've tried this once but ran into race condition issues that I didn't bother fixing and just reverted to old version.

**Comment by renderedSafe** [comment]
Yeah I suppose you could store like a second's worth of frames in a list,
and just label the frames with the time.time() they were taken, and then
the collect_data loop could just ask for the proper frame from the list.

**Comment by renderedSafe** [comment]
So, I'm now starting to think about how I'm going to stitch these 3 horizontal stripes together. Each one will have a shape of [W, (H/3), 3], how do I stitch all three of these together with numpy?

**Comment by renderedSafe** [comment]
And I did do a test to see if fewer pixels to grab does in fact equal more FPS, and yes, it does. 80 FPS grabbing 860x640

**Comment by Werseter** [comment]
Oh opencv2 just displays them as proper images, I've never thought it's an issue. What stiching is required?

**Comment by Werseter** [comment]
So you are trying to give it multiple areas to stitch. I do not know how to handle this, but from what I remember, numpy arrays are column-first ordered. Try searching up online, my first thought would be to transpose them, link the ends and transpose again. Or to feed them separately to the neural, with selected areas to limit unnecessary data from edges - for example, feeding it just the minimap, and area in about 60% center horizontally, and just until above the horizon for vertical. That's what I would do if I had the processing power ;_;

**Comment by renderedSafe** [comment]
Yeah so I have 3 threads running successfully which what I think is a
decently elegant solution, I'll upload to GitHub and share a link later.
The only problem is that the threads will start and everything, but the
script calling them to start in the first place does not seem to run
(collect_data.py) super confused

**Comment by renderedSafe** [comment]
It looks like numpy.vstack is what I need, and it's doing it, but my
collect_data isn't running it's while loop, or seemingly anything after I
initialize the threading object, so I'm taking a break to think about it.
Also, you probably know already, but you don't really need to get that
1080ti to do GPU acceleration at that level. Keras has made some recent big
progress with multi GPU support, so a couple cheap cards in parallel are
pretty formidable

**Comment by Werseter** [comment]
I'm a broke student with just a *gaming* laptop :) And I can't use any GPU acceleration with my setup - at least as far as I know. But I keep an eye out on ideas.

**Comment by renderedSafe** [comment]
I feel you. I'm looking into getting speed and other data directly from the
game so maybe I can train a network that won't require and convolutional
layers, I bet if that ends up happening training will be a lot quicker.

**Comment by renderedSafe** [comment]
Here's the 4-threading version, doesn't do much better on my machine:

`# created by Werseter

from threading import Thread
from Working.grabscreen import GrabScreen
import time
import numpy as np


class GetFrameThread:
    def __init__(self, x, y, w, h, window_title_substring):
        """
        An object that will thread through a loop, constantly updating the self.image attribute as the last value of the
        screen. Call self.start() to start the thread, and self.return_frame() to return the latest screen shot.
        :param x: farthest left part of window
        :param y: how far down the window starts. normally 40 for the window header (tested on Windows 10)
        :param w: width of window
        :param h: height of window
        :param window_title_substring: a substring of the window title
        """

        self.stop = False
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.window_title_substring = window_title_substring

        # defining the top third, middle third, and bottom third GrabScreen objects to be threaded
        self.printscreen_top = GrabScreen(window_title=self.window_title_substring,
                                      region=(self.x, self.y, self.w, int(self.h/3)))
        self.printscreen_middle = GrabScreen(window_title=self.window_title_substring,
                                      region=(self.x, int(self.h/3), self.w, int(self.h/3)))
        self.printscreen_bottom = GrabScreen(window_title=self.window_title_substring,
                                             region=(self.x, 2*(int(self.h/3)), self.w, int(self.h/3)))
        self.top_captured = False
        self.middle_captured = False
        self.bottom_captured = False

        self.image_top = None
        self.image_middle = None
        self.image_bottom = None

        self.image = None

        self.render_frame_times = []
        self.render_last_time = time.time()

    def start(self):
        """
        Starts the threading screenshot loop, as well as starting a thread
        checking to see if all the threads have pending frames, and stitching them together
        :return: self
        """
        Thread(target=self._get_frame, args=(self.printscreen_top,)).start()
        Thread(target=self._get_frame, args=(self.printscreen_middle,)).start()
        Thread(target=self._get_frame, args=(self.printscreen_bottom,)).start()
        Thread(target=self._stitch_images, args=()).start()

        return self

    def _stitch_images(self):
        """Runs on a dedicated thread to check for all frames pending and then stitches them
        """
        while True:
            if self.top_captured and self.middle_captured and self.bottom_captured:
                self.image = np.vstack((self.image_top, self.image_middle, self.image_bottom))
                self.top_captured = False
                self.middle_captured = False
                self.bottom_captured = False

                self.render_frame_times.append(time.time() - self.render_last_time)
                self.render_frame_times = self.render_frame_times[-20:]
                self.render_last_time = time.time()

    def _get_frame(self, section):
        """
        The threading loop that takes quick screenshots, this runs and constantly updates the self.image attribute
        by calling a function from grabscreen.py. This function updates self.image as the RGB image of the full sized
        screen shot at every iteration.
        :section: GrabScreen object, the idea is to be able to split up the screen between threads. i.e. one thread
        grabs the top third, one the middle, and one the bottom
        :return: none
        """
        while True:
            if self.stop:
                section.clear()
                break

            image = section.get_frame()
            if section == self.printscreen_top and not self.top_captured:
                self.image_top = image
                self.top_captured = True
            elif section == self.printscreen_middle and not self.middle_captured:
                self.image_middle = image
                self.middle_captured = True
            elif section == self.printscreen_bottom and not self.bottom_captured:
                self.image_bottom = image
                self.bottom_captured = True

    def return_frame(self):
        """
        Call this function to get the last value of self.image
        :return: self.image, array of shape [H, W, 3] channels in the order RGB
        """
        return self.image

    def stop_now(self):
        """
        Stops the object's threading loop
        :return: none
        """
        self.stop = True

    def get_fps(self):
        """
        Returns the FPS based on last 20 frames
        :return: int Frames per second (FPS)
        """
        fps = (len(self.render_frame_times) / sum(self.render_frame_times))
        return fps`

**Comment by Werseter** [comment]
Well doing the splitting and merging *inside* the getFrame itself isn't what I expected :D I'd simply initialize 3 objects. But I suppose `ifs` do not make much of an overhead.

So, you're saying the improvement isn't felt on large images, and the splitting attempt haven't helped?

**Comment by Werseter** [comment]
If we really are to push for performance, the part slicing out the alpha channel can be commented out, and the image set to hold 4-dimensional data. In the grabscreen part I don't think it can be optimized much further in this approach. I'm not sure what does the alpha channel contain as I've never bothered checking, so I do not know how the network would handle it, except getting the size up.

You also should make a handle to be able to stop the merging thread.

**Comment by renderedSafe** [comment]
It seems to be marginal, but right now I'm running just the lower two thirds, and getting about 20 FPS from the threading and 40 from the collect loop.

**Comment by renderedSafe** [comment]
I really should make a handle.

**Comment by renderedSafe** [comment]
Looking into the multiprocessing module... my head hurts. But it seems like the solution, apparently in Python we're not running more than one thread at a time when we use Thread, we have to use Process to make another interpreter.

**Comment by Werseter** [comment]
Aaah I see, I've ran a quick comparison online and indeed threading doesn't benefit from multiple cores, only helps with making non-blocking operations. That's why I've seen a big improvement in responsiveness. I suppose I have been somewhat mistaken then, but that's a move in a correct direction. Keep me updated on your progress!

**Comment by renderedSafe** [comment]
@Werseter here's the repo with my current work on the issue.
https://github.com/renderedSafe/fast-screenshots.git

**Comment by renderedSafe** [comment]
New strategy: So our outer loop can run a lot quicker than our previous screen grabbing function, but if we asynchronously run several instances of the function all at once, they will almost certainly run out of sync, staggering their outputs, and when one completes, we have it update the image variable that our data recording loop will access. It is starting to look like I'm going to have to put this loop inside the collect_data file, because it seems like, as finicky as Python multiprocessing is, anything starting a pool or Process should be protected by `if __name__ == '__main__'`.

**Comment by renderedSafe** [comment]
Well I tried the new thing, it's on the repo listed above, it works meh, I have played with the size of the worker pool and it actually seems to be faster with just one worker, which I guess half makes sense, because it should be opening up a new process, unlike threading which just stuffs the traffic in somewhere with the stream coming from the python interp.

I'm putting this on the back burner for now, this one really was a challenge, I've only seen more ambiguous errors once.

**Comment by Werseter** [comment]
I'll give it a try in free time. I'll check out your work and try to advance from it.

**Comment by renderedSafe** [comment]
Yeah so the multiprocessing is working, but once I got it working I never got around to applying it to taking slices, only whole screenshots, so that's still a possible candidate for optimizing screen grabbing.
